[ikhal] add option to disable mouse, fixes #1289

tgy opened the pull request:
This is done by passing the `handle_mouse` Boolean argument to urwid's main loop (currently unused). By default, the mouse remains enabled (see `khal.spec`). Mouse can be disabled either by setting `enable_mouse = False` in the `[default]` section of khal's config file, or by passing the `--no-mouse` flag to the `khal interactive` or `ikhal` click commands.

fixes #1289

tgy commented on .gitignore:
```diff
@@ -19,3 +19,4 @@ env/
 venv/
 .hypothesis/
 .python-version
+.dmypy.json
```
I've added this because I'm using neovim with pylsp-mypy (with the dmypy daemon, which can create this file).

WhyNotHugo replied:
No objections on this.

Note that you can include this in `~/.config/git/ignore` and git will ignore the file for all your local repositories.

tgy replied:
Ah thanks, did not know that. Good to know :)

tgy commented on khal/cli.py:
```diff
@@ -93,6 +93,12 @@ def multi_calendar_option(f):
     return d(a(f))
 
 
+def no_mouse_option(f):
```
Since I had to use this both for `ikhal` and `khal interactive`, I figured it made sense to abstract this option into its own function decorator.

tgy commented on khal/cli.py:
```diff
@@ -480,9 +486,12 @@ def isatty(_file):
+    def interactive(ctx, include_calendar, exclude_calendar, no_mouse):
         '''Interactive UI. Also launchable via `ikhal`.'''
+        if no_mouse:
+            ctx.obj['conf']['default']['enable_mouse'] = False
```
I'm overriding the config value when the `--no-mouse` flag is used. Not sure this is the best way, but seems to work.

tgy commented on khal/ui/__init__.py:
```diff
@@ -1344,7 +1344,12 @@ def emit(self, record):
     loop = urwid.MainLoop(
-        frame, palette, unhandled_input=frame.on_key_press, pop_ups=True)
+        widget=frame,
+        palette=palette,
```
I took the liberty to use kwargs instead of args, to make the urwid API call clearer

WhyNotHugo reviewed (approved):
Looks okay to me.

tgy commented:
@WhyNotHugo i fixed the tests, can you trigger the CI build (sorry was busy with work and forgot this PR)

WhyNotHugo commented on khal/cli.py:
```diff
@@ -93,6 +93,12 @@ def multi_calendar_option(f):
 
 
+def no_mouse_option(f):
+    o = click.option('--no-mouse', is_flag=True,
```
You can use a flag that accepts both positive and negative here:

```suggestion
    o = click.option('--mouse/--no-mouse', is_flag=True, default=True
```

This changes the name of this value to `mouse` and avoids a double negation when handling it below. E.g.: `--mouse` is `True` and `--no-mouse` is `False`.

tgy replied:
nice! learning something new everyday :)

WhyNotHugo commented on khal/cli.py:
```diff
@@ -93,6 +93,12 @@ def multi_calendar_option(f):
 
 
+def no_mouse_option(f):
+    o = click.option('--mouse/--no-mouse', is_flag=True, default=True
```
Missing comma after `default=True`. Doesn't run.

WhyNotHugo commented on khal/cli.py:
```diff
@@ -480,9 +486,12 @@ def isatty(_file):
+    @no_mouse_option
     @click.pass_context
-    def interactive(ctx, include_calendar, exclude_calendar):
+    def interactive(ctx, include_calendar, exclude_calendar, no_mouse):
```
I believe this flag would now be `mouse`.

WhyNotHugo commented on khal/cli.py:
```diff
@@ -494,10 +503,13 @@ def interactive(ctx, include_calendar, exclude_calendar):
+    @no_mouse_option
     @click.pass_context
-    def interactive_cli(ctx, config, include_calendar, exclude_calendar):
+    def interactive_cli(ctx, config, include_calendar, exclude_calendar, no_mouse):
```
Ditto

tgy commented:
@WhyNotHugo addressed your comments (sorry for not running the tests before), ran the tests locally and everything passes. i think you can trigger the CI again!

geier commented on khal/cli.py:
```diff
@@ -93,6 +93,16 @@ def multi_calendar_option(f):
+    o = click.option(
+        '--mouse/--no-mouse',
+        is_flag=True,
+        default=True,
```
I believe we need to set the default to `None` and then check explicity in `interactive_cli()` and `interactive()` if `mouse` is set to something other than `None` and only override ctx if that is the case.

geier commented on khal/cli.py:
```diff
@@ -494,10 +506,12 @@ def interactive(ctx, include_calendar, exclude_calendar):
+    def interactive_cli(ctx, config, include_calendar, exclude_calendar, mouse):
         '''Interactive UI. Also launchable via `khal interactive`.'''
         prepare_context(ctx, config)
+        ctx.obj['conf']['default']['enable_mouse'] = mouse
```
Something like
```python
if mouse is not None:
        ctx.obj['conf']['default']['enable_mouse'] = mouse
```

geier reviewed (changes requested):
Thanks for the PR, looks good. But I believe setting the default of the click flag to True messes with overriding the config variable. See comments.

tgy commented:
@geier done! thanks for catching that

geier commented:
@tgy I don't see any changes, did you push?

geier commented:
Works, thanks!
